### ISSUE TITLE
midi.translate.midiEventsToInstrument is totally broken!

### ISSUE THREAD

**Issue by asudakov** [cuthbertLab/music21#1518]
**music21 version**

5.5.0

**Problem summary**
midi.translate.midiEventsToInstrument

**Steps to reproduce**

```
for i in range (0, 128):
  me = midi.MidiEvent(midi.MidiTrack(0))
  me.type = 0xC0 #midi.ChannelVoiceMessages.PROGRAM_CHANGE
  me.data = i  # MIDI program 54: Voice Oohs
  inst = midi.translate.midiEventsToInstrument(me)
  print('Midi', i, inst.instrumentSound)
```

**Expected vs. actual behavior**
Compare to General MIDI [https://en.wikipedia.org/wiki/General_MIDI](https://en.wikipedia.org/wiki/General_MIDI)
**More information**

**Comment by jacobtylerwalls** [comment]
Yes, this is a known missing feature indicated with these TODO statements:

https://github.com/cuthbertLab/music21/blob/aa7b06836a78096220a977a85c9ecd6e368fcec4/music21/instrument.py#L1928-L1930

https://github.com/cuthbertLab/music21/blob/aa7b06836a78096220a977a85c9ecd6e368fcec4/music21/instrument.py#L2086-L2088

Would you like to prepare a patch?

**Comment by asudakov** [comment]
Is another mapping really necessary? Instrument classes already define everything including "instrumentSound".

After poking at 8.1.0 I believe music21.converter does parse out program_change events from MIDI. I believe there is a bug in track.getInstrument() or how track is being initialized when parsing from midi.

With some midis that I tried track.getInstrument() gives me un-initialized Instrument object, while  track.getInstrument()[0] gives me proper instrument subclass.

**Comment by jacobtylerwalls** [comment]
> Is another mapping really necessary?

If you're asking if the midi program number is preserved, yes, that's already stored on .midiProgram. The TODOs are for setting instrumentSound also. In terms of coalescing all the sampler sounds to a KeyboardInstrument, that's a feature; there won't be 128 instrument subclasses to provide a 1-1 map with MIDI programs.

**Comment by jacobtylerwalls** [comment]
> With some midis that I tried track.getInstrument() gives me un-initialized Instrument object, while track.getInstruments()[0] gives me proper instrument subclass.

There's an asymmetry between the default values of the `recurse` keyword to those two methods. That would explain the discrepancy. If you want to suggest setting `recurse=True` as default for `getInstrument()` to match `getInstruments()`, that should be discussed on a separate issue.

**Comment by asudakov** [comment]
What I am trying to say I am getting pretty reliably (channel 10 drum track apart and some odd cases) instrument set with program_change event in MIDI file as track.getInstruments()[0]. Yet, rack.getInstrument() is pretty much always "blank" Instrument object.

There gotta be a bug 😉 how getInstrument() is setup when parsing MIDI vs. getInstruments() . Shouldn't getInstrument() be equal to getInstruments()[0]? 

Also, just FYI, I patched some MIDI to have "instrument_name" MetaEvent to exploit "MuseScore hack" and use instrument.fromString route...
`
      instrument_name = music21.instrument.MIDI_PROGRAM_TO_INSTRUMENT[program_change_events[0].program]().instrumentName
      meta_message = MetaMessage('instrument_name', name=instrument_name, time=0)
      track.insert(0, meta_message)
`
Seems to make no difference. Similarly, I have getInstruments() populated with exact same Instruments as program_change files while getInstrument() is "blank" Instrument object.


Finally, what about 'em drums? Why cannot I get instrument[s] for them with channel change to channel 10?

**Comment by asudakov** [comment]
> > Is another mapping really necessary?
> 
> If you're asking if the midi program number is preserved, yes, that's already stored on .midiProgram. The TODOs are for setting instrumentSound also. In terms of coalescing all the sampler sounds to a KeyboardInstrument, that's a feature; there won't be 128 instrument subclasses to provide a 1-1 map with MIDI programs.

Me thinks - why bother with another mapping? MIDI_PROGRAM_TO_INSTRUMENT is pretty much covers GM as is 😉...

Me thinks, it is just getInstrument() vs. getInstrument**s**() when parsing MIDI 😉

**Comment by jacobtylerwalls** [comment]
> Shouldn't getInstrument() be equal to getInstruments()[0]?

Not if you're providing opposite (default) arguments for the recurse keyword. See some recent mailing list posts for an intro to music21's "lists of lists" ontology.

**Comment by mscuthbert** [comment]
Can you update to latest music21 (release or master) and then continue the issue?  There's not the resources to consider work on such an old version.  Thank you!

**Comment by asudakov** [comment]
I would recommend closing this issue. 8.1.0 is good at properly getting midi instruments from MIDI program change events as getInstruments()[0] vs. getInstrument().

Perhaps, some refactoring to make getInstrument() check getInstruments()[0] would be in order? Or perhaps setting getInstruments()[0] when parsing midi would be better way to go?
